RFC37: ckb2021 overview (#242)
nervosnetwork/rfcs: rfcs/0037-ckb2021/0037-ckb2021.md

Author: Ian Yang
Organization: Nervos Foundation
Created: 2021-07-24
---

# CKB2021 Overview

The blockchain CKB is introducing a hard fork which will be complete by October, 2021. CKB2021 Overview outlines what you need to know for this upgrade.

This is the first hard fork planned when CKB launches. The idea of a hard fork is to update the rules that are set to the network. The changes are not backward compatible. CKB2021 refers to the new CKB edition after the hard fork, where the old edition before the fork is CKB2019.

## What's in CKB2021

CKB2021 brings many new features and bug fixes. As usual, these changes are proposed via RFCs. The appendix has a list of all the RFCs related to CKB2021.

The upgrade is divided into three categories. 

First, CKB VM gets a major upgrade. CKB2021 will bundle CKB VM v1, in addition to the v0 in CKB2019. Scripts will be executed on v1 unless users opt in to use v0 by setting the script hash type to `data`.  

Second, CKB2021 adds a new field `extension` in the block. This is reserved for future upgrades such as flyclient.

Lastly, there are a bunch of consensus patches to fix bugs and make the consensus rules reasonable.

### CKB VM v1

> Since CKB2021, there will be multiple VM versions available. ﻿[RFC32] designs a mechanism to choose the CKB VM version. It uses the `hash_type` field in the Script structure.

Reviewer: ```suggestion
Since CKB2021, there will be multiple VM versions available. ﻿[RFC32] introduces a CKB VM version mechanism. It piggybacks on the `hash_type` field in the Script structure.
```

Although the field is added to the block body, nodes can synchronize the block header and this field together without overhead.

CKB2021 will not parse and verify the field after the activation. Instead, it enables a future soft fork to give the definition of the extension field. For example, flyclient can store the hash in the extension field.

### Consensus Patches

[RFC28] uses block timestamp as the start time for the relative timestamp `since` field, instead of the median of previous 37 blocks. This simplifies the `since` maturity calculation.

[RFC29] allows multiple matches on dep cells via type script hash when these cells have the same data. It removes unnecessary restrictions when there's no ambiguity to choose matched script code.

[RFC30] ensures that the index is less than the length in the `since` field using epoch as the time measure. It avoids the ambiguity because of the inconsistent behavior when using relative and absolute epoch `since`.

> [RFC36] removes header deps immature rule. Now developers can choose how long to wait until a header can be used as a dep header.

Reviewer: ```suggestion
[RFC36] removes header deps immature rule, allowing developers to choose how long to wait until a header can be used as a dep header.
```

## CKB2021 Timeline

> The mainnet upgrade will be divided into three phases.

Reviewer: ```suggestion
The mainnet upgrade is divided into three phases.
```

> * **Stage 1**: An RC version of 0.100.0 is ready for preview on July 16 via nervosnetwork/ckb [releases](https://github.com/nervosnetwork/ckb/releases). It will introduce the incompatible changes to help developers to adapt their tools and apps to CKB2021. But this version does not activate the consensus incompatible changes in CKB2021. Developers can try them by running a dev chain locally.

Reviewer: ```suggestion
* **Stage 1**: An RC version of 0.100.0 is ready for preview on July 16 2021 via nervosnetwork/ckb [releases](https://github.com/nervosnetwork/ckb/releases). It will introduce the incompatible changes to help developers to adapt their tools and apps to CKB2021. But this version does not activate the consensus incompatible changes in CKB2021. Developers can test the new rules by running a dev chain locally.
```

> * **Stage 2** (Estimated in August): An updated RC version will activate CKB2021 on Aggron testnet about 1 week after the release.

Reviewer: Date needs update.

> * **Stage 3** (Estimated in September): The final release 0.100.0 will activate CKB2021 on Lina mainnet about 4 weeks after the release.

Reviewer: Date needs update.

> * **Stage 3** (Estimated in September): The final release 0.100.0 will activate CKB2021 on Lina mainnet about 4 weeks after the release.

Author: ```suggestion
* **Stage 3** (Estimated in March 2022): The final release 0.102.0 will activate CKB2021 on Lina mainnet about 4 weeks after the release.
```

## Upgrade Strategies

> First, the SDK, Tool, and dApps authors must adapt to any 1.100.0 rc version.

Reviewer 2: Might be `0.100.0 rc version.`

Author: ```suggestion
First, the SDK, Tool, and dApps authors must adapt to any 0.100.0 rc version.
```

Reviewer: Seems there's no "Second, ..." in the paragraphs below.

> There are two strategies to upgrade to the CKB2021 consensus.

Reviewer: Are these two strategies for the CKB team or ecosystem developers? How to choose between the two?

> There are two strategies to upgrade to the CKB2021 consensus.

Author: ```suggestion
There are two strategies for ecosystem developers to upgrade to the CKB2021 consensus, depending whether the developers can pause the app during the fork activation.
```

> - Release two different versions or use the feature switcher. Manually deploy the newer version or enable the feature CKB2022 after the fork activation.

Reviewer: ```suggestion
- Release two different versions or use the feature switcher. Manually deploy the newer version or enable the feature CKB2021 after the fork activation.
```
